### ISSUE TITLE
Small offset in detection and bad practice

### ISSUE THREAD

**Issue by Juanjojr9** [ultralytics/yolov5#6554]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello

I have been training the Yolov5x6 model on my own dataset. I want to detect the pictures class, as COCO does not have this class.
Once the network is trained, I use detect.py. But I have a small problem, there is a small offset in the detection.

![image](https://user-images.githubusercontent.com/56088039/152785136-77b59ed7-c25a-437b-87ee-8f87f02d8ddf.png)
![image](https://user-images.githubusercontent.com/56088039/152785389-59e08b7c-e5a1-40d2-8336-fe3e21ce057c.png)

How could I solve it?

Another question I have.
As I had no CUDA memory, I reduced the batch size to 1. I know it's a bad practice, but I don't understand why. What is the reason for this?

What would be a best practice?

Thank you very much, I look forward to receiving your answers
Regards
JJ

**Comment by glenn-jocher** [comment]
@Juanjojr9 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by Juanjojr9** [comment]
@glenn-jocher Thank you very much for your answer. I will try to see how to improve it.

Regarding the batch size question, why does it have to be as big as possible ? I have batch size to 1 and I don't know if it is the right thing to do.

What should I do?

**Comment by glenn-jocher** [comment]
@Juanjojr9 👋 Hello! Thanks for asking about **CUDA memory issues**. Batch size should be as large as possible for improved BatchNorm2d() statistics aggregation during training.

YOLOv5 🚀 can be trained on CPU, [single-GPU](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data), or [multi-GPU](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training). When training on GPU it is important to keep your batch-size small enough that you do not use all of your GPU memory, otherwise you will see a CUDA Out Of Memory (OOM) Error and your training will crash. You can observe your CUDA memory utilization using either the `nvidia-smi` command or by viewing your console output:

<img width="800" alt="Screenshot 2021-05-28 at 12 19 51" src="https://user-images.githubusercontent.com/26833433/131122318-8141d1ce-13de-44bf-bbdc-487bad442ff0.png">

### CUDA Out of Memory Solutions

If you encounter a CUDA OOM error, the steps you can take to reduce your memory usage are:

- Reduce `--batch-size`
- Reduce `--img-size`
- Reduce model size, i.e. from YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s
- Train with [multi-GPU](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) at the same `--batch-size`
- Upgrade your hardware to a larger GPU
- Train on free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

### AutoBatch

You can use [YOLOv5 AutoBatch](https://github.com/ultralytics/yolov5/pull/5092) (NEW) to find the best batch size for your training by passing `--batch-size -1`. AutoBatch will solve for a 90% CUDA memory-utilization batch-size given your training settings. AutoBatch is experimental, and only works for Single-GPU training. It may not work on all systems, and is not recommended for production use.

<img width="1746" alt="Screenshot 2021-11-06 at 12 31 10" src="https://user-images.githubusercontent.com/26833433/140608327-9c0c0baf-a0ee-4551-a4ff-5ee26b155c51.png">

Good luck 🍀 and let us know if you have any other questions!

**Comment by Juanjojr9** [comment]
@glenn-jocher Thank you for your quick response.
I am using Google Colab. These are my parameters

![image](https://user-images.githubusercontent.com/56088039/152834581-492be53b-8d61-46c7-ba29-c91dae6a5c2c.png)

As I ran out of CUDA memory I decided to lower the batch size.

Would it be better to lower the image size and try to increase the batch size? Or use a smaller model?

Or what would be the best practice?

**Comment by glenn-jocher** [comment]
@Juanjojr9 up to you, see existing response https://github.com/ultralytics/yolov5/issues/6554#issuecomment-1031685830

**Comment by Juanjojr9** [comment]
@glenn-jocher I have seen it, but what I meant by my previous question was whether it is more convenient to reduce the batch size to the maximum, without touching the image size, or to reduce both a bit.

**Comment by glenn-jocher** [comment]
@Juanjojr9 depends on use case, available resources, organization priorities etc. Use above guide and follow recommendations.

**Comment by Juanjojr9** [comment]
@glenn-jocher  OK, perfect.

Thank you very much for all your help. It is very useful for me

Regards
